src/OrchardCore/OrchardCore.Environment.Shell.Abstractions/Builders/ShellContext.cs

```diff
@@ -58,75 +65,98 @@ public IServiceScope EnterServiceScope()
         /// </summary>
         public int ActiveRequests => _refCount;
 
-        public void RequestStarted()
-        {
-            Interlocked.Increment(ref _refCount);
-        }
-
-        public void RequestEnded()
-        {
-            var refCount = Interlocked.Decrement(ref _refCount);
-        }
-
-        public bool CanTerminate => _released && _refCount == 0;
+        
+        public bool CanTerminate => _released && _refCount == 0 && !_disposed;
 
         /// <summary>
         /// Mark the <see cref="ShellContext"/> has a candidate to be released.
         /// </summary>
         public void Release()
         {
+            if (_released == true)
+            {
+                // Prevent infinite loops with circular dependencies
+                return;
+            }
+
             // When a tenant is changed and should be restarted, its shell context is replaced with a new one, 
             // so that new request can't use it anymore. However some existing request might still be running and try to 
             // resolve or use its services. We then call this method to count the remaining references and dispose it 
             // when the number reached zero.
 
             _released = true;
-        }
 
-        public void Dispose()
-        {
-            Dispose(true);
-            GC.SuppressFinalize(this);
-        }
-
-        protected virtual void Dispose(bool disposing)
-        {
-            if (!_disposed)
+            lock (this)
             {
-                if (disposing)
+                if (_dependents == null)
+                {
+                    return;
+                }
+
+                foreach (var dependent in _dependents)
                 {
+                    dependent.Release();
                 }
```

Comment:
So, when you release a shell, its dependents are released, but i didn't find the code path where you 1st release this shell.

Reply:
`ShellHost`

Reply:
Oh yes, as before, i'm tired ;)